### PR TITLE
Add vsock test

### PR DESCRIPTION
**Pull request by chavafg**
- Add support for the CI to be able to configure kata-runtime with `use_vsock = true`
- Add test to verify that kata-proxy does not run when using vsock.

### CODE REVIEW

**Comment by chavafg** [comment]
I have added a new job `jenkins-ci-fedora-27-vsock` that runs all tests using vsocks, when this gets merged, I will add another job for the runtime repository.

File: .ci/install_runtime.sh [hunk 1]
```diff
@@ -56,6 +58,20 @@ else
 	sudo sed -i -e '/^initrd =/d' ${runtime_config_path}
 fi
 
+if [ "$USE_VSOCK" == "yes" ]; then
+	echo "Configure use of VSOCK in ${runtime_config_path}"
+	sudo sed -i -e 's/^#use_vsock.*/use_vsock = true/' "${runtime_config_path}"
+
+	vsock_module="vhost_vsock"
+	echo "Check if ${vsock_module} is loaded"
+	if lsmod | grep "$vsock_module" &> /dev/null ; then
```
Comment (devimc) [inline .ci/install_runtime.sh]: grep -q "$vsock_module"

File: integration/docker/vsock_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+				Skip("Use of vsock not enabled")
+			}
+			if val != "yes" {
+				Skip("Use of vsock not enabled")
+			}
+			args = []string{"--name", id, "-d", Image, "top"}
+			_, _, exitCode = dockerRun(args...)
+			Expect(exitCode).To(Equal(0))
+
+			// Check no kata-proxy process is running
```
Comment (devimc) [inline integration/docker/vsock_test.go]: Do not use `ps`, use https://github.com/kata-containers/tests/blob/master/process.go#L66 instead

File: integration/docker/vsock_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	)
+
+	AfterEach(func() {
+		Expect(RemoveDockerContainer(id)).To(BeTrue())
+		Expect(ExistDockerContainer(id)).NotTo(BeTrue())
+	})
+
+	Context("when using vsock", func() {
+		It("should not create a kata-proxy process", func() {
+			val, ok := os.LookupEnv("USE_VSOCK")
```
Comment (devimc) [inline integration/docker/vsock_test.go]: Read configuration file instead https://github.com/kata-containers/tests/blob/master/process.go#L67

btw you have to update https://github.com/kata-containers/tests/blob/master/config.go#L26 and add `use_vsock`

**Comment by jodh-intel** [comment]
lgtm

File: integration/docker/docker.go [hunk 4]
```diff
@@ -64,6 +64,18 @@ func runDockerCommandWithPipe(stdin *bytes.Buffer, command string, args ...strin
 	return runDockerCommandWithTimeoutAndPipe(stdin, time.Duration(tests.Timeout), command, args...)
 }
 
+// GetContainerID returns the docker container ID
+func GetContainerID(name string) (string, error) {
```
Comment (devimc) [inline integration/docker/docker.go]: you don't need this function, see `dockerInspect`

File: integration/docker/vsock_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+			useVsock := KataConfig.Hypervisor[DefaultHypervisor].Vsock
+			if !useVsock {
+				Skip("Use of vsock not enabled")
+			}
+			args = []string{"--name", name, "-d", Image, "top"}
+			_, _, exitCode = dockerRun(args...)
+			Expect(exitCode).To(Equal(0))
+
+			inspectArgs := []string{"--format", "{{.Id}}", name}
+			ctrID, _, exitCode := dockerInspect(inspectArgs...)
```
Comment (devimc) [inline integration/docker/vsock_test.go]: `dockerInspect("--format", "{{.Id}}", name)`

File: integration/docker/vsock_test.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+			Expect(exitCode).To(Equal(0))
+
+			// Check no kata-proxy process is running
+			isProxyRunning := ProxyRunning(ctrID)
+			Expect(isProxyRunning).To(Equal(false))
```
Comment (devimc) [inline integration/docker/vsock_test.go]: `Expect(ProxyRunning(ctrID)).To(BeFalse())`

File: integration/docker/vsock_test.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+	AfterEach(func() {
+		Expect(RemoveDockerContainer(name)).To(BeTrue())
+		Expect(ExistDockerContainer(name)).NotTo(BeTrue())
+	})
+
+	Context("when using vsock", func() {
+		It("should not create a kata-proxy process", func() {
+			useVsock := KataConfig.Hypervisor[DefaultHypervisor].Vsock
+			if !useVsock {
```
Comment (devimc) [inline integration/docker/vsock_test.go]: `if !KataConfig.Hypervisor[DefaultHypervisor].Vsock {`

File: integration/docker/vsock_test.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	. "github.com/kata-containers/tests"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("vsock test", func() {
+	var (
+		args     []string
+		name     string = randomDockerName()
```
Comment (devimc) [inline integration/docker/vsock_test.go]: this initialization should be in `BeforeEach`

File: integration/docker/vsock_test.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+			args = []string{"--name", name, "-d", Image, "top"}
+			_, _, exitCode = dockerRun(args...)
+			Expect(exitCode).To(Equal(0))
+
+			ctrID, _, exitCode := dockerInspect("--format", "{{.Id}}", name)
+			Expect(exitCode).To(Equal(0))
+
+			// Check no kata-proxy process is running
+			isProxyRunning := ProxyRunning(ctrID)
+			Expect(isProxyRunning).To(BeFalse())
```
Comment (devimc) [inline integration/docker/vsock_test.go]: `Expect(ProxyRunning(ctrID)).To(BeFalse())`

File: integration/docker/vsock_test.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	})
+
+	Context("when using vsock", func() {
+		It("should not create a kata-proxy process", func() {
+			if !KataConfig.Hypervisor[DefaultHypervisor].Vsock {
+				Skip("Use of vsock not enabled")
+			}
+			args = []string{"--name", name, "-d", Image, "top"}
+			_, _, exitCode = dockerRun(args...)
+			Expect(exitCode).To(Equal(0))
```
Comment (devimc) [inline integration/docker/vsock_test.go]: `Expect(exitCode).To(BeZero())`

File: integration/docker/vsock_test.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+			ctrID, _, exitCode := dockerInspect("--format", "{{.Id}}", name)
+			Expect(exitCode).To(Equal(0))
```
Comment (devimc) [inline integration/docker/vsock_test.go]: ``Expect(exitCode).To(BeZero())``

**Comment by chavafg** [comment]
I see that frequently I got this error when running using vsocks:
```
Failed to check if grpc server is working: context deadline exceeded: unknown.
```

Opened: https://github.com/kata-containers/runtime/issues/576

any idea @jcvenegas @devimc @sboeuf

**Comment by devimc** [comment]
@chavafg  I can't reproduce it

**Comment by devimc** [comment]
@chavafg  
>  "it works on my machine"

**Comment by amshinde** [comment]
lgtm.
@devimc can you take a look at the issue @chavafg pointed out.
@chavafg Were you able to reproduce the issue on a Azure instance?

**Comment by chavafg** [comment]
@amshinde, not on azure, but I could reproduce using a local VM. Also @nitkon faced this same issue: https://github.com/kata-containers/runtime/issues/576#issuecomment-412805705

**Comment by amshinde** [comment]
@chavafg Can you provide @devimc the exact setup steps or a setup you have available? 
@devimc We have seen reports from other users as well.

**Comment by devimc** [comment]
@amshinde  can you show me those reports ?

**Comment by chavafg** [comment]
@devimc opened https://github.com/kata-containers/agent/pull/323 to solve this.

**Comment by devimc** [comment]
@amshinde in the issue that you mention (kata-containers/kata-containers#28) vsocks are not being used 

```

[Hypervisor]
  MachineType = "pc"
  Version = "QEMU emulator version 2.11.0 (v2.11.0-rc0-308-ga39e0b3-dirty)\nCopyright (c) 2003-2017 Fabrice Bellard and the QEMU Project developers"
  Path = "/usr/bin/qemu-lite-system-x86_64"
  BlockDeviceDriver = "virtio-scsi"
  Msize9p = 8192
  Debug = false
  UseVSock = false
```

**Comment by chavafg** [comment]
added dnm label as vsock job still in progress.

http://jenkins.katacontainers.io/job/kata-containers-tests-vsock-fedora-27-PR/

**Pull request merged**